### PR TITLE
Optimization - avoid temporary list objects, unnecessary function call

### PR DESCRIPTION
**Pull request by mlyundin**
Please review.

### CODE REVIEW

File: scrapy/crawler.py [hunk 1]
```diff
@@ -173,7 +173,7 @@ def stop(self):
 
         Returns a deferred that is fired when they all have ended.
         """
-        return defer.DeferredList([c.stop() for c in list(self.crawlers)])
+        return defer.DeferredList(c.stop() for c in list(self.crawlers))
```
Comment (dangra) [inline scrapy/crawler.py]: Old twisted versions doesn't accept a generator as DeferredList argument.

btw, why the list(self.crawlers) call?

Comment (mlyundin) [inline scrapy/crawler.py]: Ok. Thank you.

About crawlers. I can only suppose that it could be non iterable object...

Comment (dangra) [inline scrapy/crawler.py]: but if it is not iterable it will fail to build a list too

Comment (mlyundin) [inline scrapy/crawler.py]: I agree, That is why I gave updated my comment before)

File: scrapy/utils/spider.py [hunk 2]
```diff
@@ -4,15 +4,11 @@
 import six
 
 from scrapy.spiders import Spider
-from scrapy.utils.misc import  arg_to_iter
+from scrapy.utils.misc import arg_to_iter as iterate_spider_output
```
Comment (kmike) [inline scrapy/utils/spider.py]: if we go this way it worths adding a comment about why is there an unused import

Comment (mlyundin) [inline scrapy/utils/spider.py]: Done.

File: scrapy/xlib/tx/client.py [hunk 3]
```diff
@@ -975,7 +975,7 @@ class ContentDecoderAgent(object):
     def __init__(self, agent, decoders):
         self._agent = agent
         self._decoders = dict(decoders)
-        self._supported = ','.join([decoder[0] for decoder in decoders])
+        self._supported = ','.join(decoder[0] for decoder in decoders)
```
Comment (dangra) [inline scrapy/xlib/tx/client.py]: do not modify scrapy.xlib.client code, it is a verbatim copy of twisted sources for the specific case of supporting http1.1 in old twisted version.

We are more close to remove this code than improving it.

File: scrapy/utils/spider.py [hunk 4]
```diff
@@ -4,15 +4,12 @@
...
 from scrapy.spiders import Spider
-from scrapy.utils.misc import  arg_to_iter
+# Let put function here and give more meaningful name in context in which it will be used
+from scrapy.utils.misc import arg_to_iter as iterate_spider_output
 
 logger = logging.getLogger(__name__)
 
 
-def iterate_spider_output(result):
-    return arg_to_iter(result)
```
Comment (dangra) [inline scrapy/utils/spider.py]: I think the question is why do we maintain this alias and do not deprecate it instead.

File: scrapy/xlib/tx/client.py (hunk 3 above)
Comment (mlyundin) [inline scrapy/xlib/tx/client.py]: Done.

File: scrapy/utils/spider.py (hunk 4 above)
Comment (mlyundin) [inline scrapy/utils/spider.py]: So What should we do?

Comment (dangra) [inline scrapy/utils/spider.py]: leave this change out of this PR so we can merge and discuss in another PR :)

**Comment by nramirezuy** [comment]
@mlyundin can you rebase to master?

**Comment by mlyundin** [comment]
I suppose I have already rebased my branch to master

$ git rebase upstream/master
Current branch develop is up to date.

**Comment by mlyundin** [comment]
Review hung. Is there action required from my side?

**Comment by nramirezuy** [comment]
@kmike @dangra are we missing tests for the code changed?

**Comment by dangra** [comment]
@nramirezuy I don't think so. +1 to merge

**Comment by kmike** [comment]
Well yes, I have two reservations for this PR. The first is that some of them are unnecessary, and they don't optimize anything: generator expressions are slower than list comprehensions, and memory savings in most places where code is changed are not important (slowdown is not important either though). It is not like anyone has 100000 spider contracts or 10000-segment spider module names. 

Second is that the PR makes changes to untested parts of code (coverage reports shows this). For example, if ContractsManager would have required a list (like DeferredList did in old Twisted versions) we may not notice it.

That said, I agree with some of the cleanups; they are not optimizations, but they remove unnecessary braces in source code. So +1 to merge it if either changes to untested code are removed or new tests are added.

File: scrapy/commands/genspider.py [hunk 5]
```diff
@@ -88,8 +88,8 @@ def _genspider(self, module, name, domain, template_name, template_file):
             'module': module,
             'name': name,
             'domain': domain,
-            'classname': '%sSpider' % ''.join([s.capitalize() \
-                for s in module.split('_')])
+            'classname': '%sSpider' % ''.join(s.capitalize() \
```
Comment (dangra) [inline scrapy/commands/genspider.py]: backslash is not needed

**Comment by dangra** [comment]
@kmike comments are fair but I think it is more clean code for same functionality when performance is not measured or taken into consideration at all.

**Comment by kmike** [comment]
Yeah, I also like the look of the updated code more.

**Comment by nramirezuy** [comment]
There are 2 lines untested; we can strip those 2 and merge the rest, create a new PR with these 2 lines?

**Comment by dangra** [comment]
@nramirezuy I don't think that is needed, the untested looks fine.

**Comment by kmike** [comment]
Ok, let's just merge this :) Thanks @mlyundin for the patience!

**Comment by dangra** [comment]
enough :bike: :circus_tent: !

**Comment by nramirezuy** [comment]
:bear:

**Pull request merged**